src/main/java/com/avispl/symphony/dal/device/core510i/QRCCommunicator.java

```diff
@@ -0,0 +1,440 @@
+			this.socket = new Socket(this.host, this.port);
+			this.socket.setTcpNoDelay(true);
+			this.socket.setKeepAlive(true);
+			this.socket.setOOBInline(true);
+			this.socket.setSoTimeout(socketTimeout);
+		}
+	}
+
+	/**
+	 * This method is used to destroy a channel actually destroy a socket
+	 */
+	public void destroyChannel() {
+		if (null != this.socket) {
+			try {
+				if (this.socket.isConnected()) {
+					this.socket.close();
+				}
+			} catch (Exception var2) {
+				if (this.logger.isWarnEnabled()) {
+					this.logger.warn("error seen on destroyChannel", var2);
+				}
+			}
+
+			this.socket = null;
+		}
+	}
+
+	/**
+	 * This method is used to send a JSON RPC to a device
+	 *
+	 * @param data This is the data to be sent
+	 * @return String[] This returns the reply received from the device.
+	 */
+	protected String[] send(String data) throws Exception {
+		if (!this.isInitialized()) {
+			throw new IllegalStateException("QRCCommunicator cannot be used before init() is called");
+		}
+
+		if (null == data) {
+			throw new IllegalArgumentException("Send data is null");
+		}
+
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Sending command: " + data + " to: " + this.host + " port: " + this.port);
+		}
+
+		Lock writeLock = this.lock.writeLock();
+		writeLock.lock();
+
+		String[] response;
+		try {
+			response = this.send(data, true);
+		} finally {
+			this.destroyChannel();
+			writeLock.unlock();
+		}
+
+		return response;
+	}
+
+	/**
+	 * This method is used to send a JSON RPC to a device
+	 *
+	 * @param data This is the data to be sent
+	 * @param retryOnError This is the flag to retry sending data to device when error
+	 * @return String[] This returns the reply received from the device.
+	 */
+	private String[] send(String data, boolean retryOnError) throws Exception {
+		try {
+			if (!this.isChannelConnected()) {
+				this.createChannel();
+				this.status.setLastTimestamp(System.currentTimeMillis());
+				this.status.setConnectionState(ConnectionState.Connected);
+				this.status.setLastError(null);
+			}
+
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Sending: " + data + " to: " + this.host + " port: " + this.port);
+			}
+
+			String[] response = this.internalSend(data);
+
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Received response: " + Arrays.toString(response) + " from: " + this.host + " port: " + this.port);
+			}
+
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("Received response: " + Arrays.toString(response) + " from: " + this.host + " port: " + this.port);
+			}
+
+			this.status.setLastTimestamp(System.currentTimeMillis());
+			return response;
+		} catch (CommandFailureException ex1) {
+			if (this.logger.isErrorEnabled()) {
+				this.logger.error("Command failed " + data + " to: " + this.host + " port: " + this.port + " connection state: " + this.status.getConnectionState(), ex1);
```

Comment:
Maybe string template or stringbuilder (or string.concat) would be easier to read here (and other similar error/debug messages)?

Reply:
ok thank you, I will fix it